python-checks/src/main/java/org/sonar/python/checks/ConfusingTypeCheckingCheck.java

```diff
@@ -110,7 +112,21 @@ public boolean isValidSubscription(Expression subscriptionObject, String require
         // handled by S5644
         return true;
       }
-      return type.declaresMember(requiredMethod);
+      return isUsedInsideInExpression(subscriptionObject) || type.declaresMember(requiredMethod);
+    }
+
+    private static boolean isUsedInsideInExpression(Expression subscriptionObject) {
+      return TreeUtils.getSymbolFromTree(subscriptionObject)
+        .filter(symbol -> symbol.usages().stream().anyMatch(usage -> isRightOperandInExpression(usage.tree())))
+        .isPresent();
+    }
+
+    private static boolean isRightOperandInExpression(Tree tree) {
+      Tree parent = tree.parent();
+      if (parent instanceof InExpression) {
```

Comment:
_Minor_: I think you can use `Tree::is` here instead.

Reply:
If you don't mind, I'd prefer to use `instanceof` here so that it will also work in case of parent being `null`